components/scream/src/physics/rrtmgp/atmosphere_radiation.cpp

```diff
@@ -590,23 +609,65 @@ void RRTMGPRadiation::run_impl (const int dt) {
     // set_vmr requires the input array size to have the correct size,
     // and the last chunk may have less columns, so create a temp of
     // correct size that uses m_buffer.tmp2d's pointer
+    //
+    // h2o is taken from qv and requies no initialization here;
+    // o3 is computed elsewhere (either read from file or computed by chemistry);
+    // n2 and co are set to constants and are not handled by trcmix;
+    // the rest are handled by trcmix
     real2d tmp2d = subview_2d(m_buffer.tmp2d);
+    const auto gas_mol_weights = m_gas_mol_weights;
     for (int igas = 0; igas < m_ngas; igas++) {
       auto name = m_gas_names[igas];
-      auto fm_name = name=="h2o" ? "qv" : name;
-      auto d_temp  = get_field_in(fm_name).get_view<const Real**>();
-      const auto gas_mol_weights = m_gas_mol_weights;
+      auto d_vmr = get_field_out(name + "_volume_mix_ratio").get_view<Real**>();
+      if (name == "h2o") {
+        // h2o is (wet) mass mixing ratio in FM, otherwise known as "qv", which we've already read in above
+        // Convert to vmr
+        const auto policy = ekat::ExeSpaceUtils<ExeSpace>::get_default_team_policy(ncol, m_nlay);
+        Kokkos::parallel_for(policy, KOKKOS_LAMBDA(const MemberType& team) {
+          const int i = team.league_rank();
+          const int icol = i + beg;
+          Kokkos::parallel_for(Kokkos::TeamThreadRange(team, nlay), [&] (const int& k) {
+            d_vmr(icol,k) = PF::calculate_vmr_from_mmr(gas_mol_weights[igas],d_qv(icol,k),d_qv(icol,k));
+          });
+        });
+        Kokkos::fence();
+      } else if (name == "o3") {
+        // We read o3 in as a vmr already
+      } else if (name == "n2") {
+        // n2 prescribed as a constant value
+        Kokkos::deep_copy(d_vmr, m_params.get<Real>("n2vmr", 0.7906));
+      } else if (name == "co") {
+        // co prescribed as a constant value
+        Kokkos::deep_copy(d_vmr, m_params.get<Real>("covmr", 1.0e-7));
+      } else {
+        // This gives (dry) mass mixing ratios
+        scream::physics::trcmix(
+          name, m_lat, d_pmid, d_vmr,
+          m_co2vmr_rad, m_co2vmr, m_n2ovmr, m_ch4vmr, m_f11vmr, m_f12vmr
+        );
+        // Back out volume mixing ratios
+        const auto air_mol_weight = PC::MWdry;
+        const auto policy = ekat::ExeSpaceUtils<ExeSpace>::get_default_team_policy(m_ncol, m_nlay);
+        Kokkos::parallel_for(policy, KOKKOS_LAMBDA(const MemberType& team) {
+          const int i = team.league_rank();
+          Kokkos::parallel_for(Kokkos::TeamThreadRange(team, m_nlay), [&] (const int& k) {
```

Comment:
`m_nlay` is an implicit use of `this`. You should use `nlay`.

Reply:
Ah yeah you're right. Will fix, thanks for finding.

Reply:
Fixed.